src/react/elements.js

```diff
@@ -15,38 +15,158 @@ import {
   AbstractValue,
   AbstractObjectValue,
   ArrayValue,
+  FunctionValue,
   NullValue,
   NumberValue,
   ObjectValue,
+  StringValue,
   Value,
 } from "../values/index.js";
 import { Create } from "../singletons.js";
 import invariant from "../invariant.js";
 import { Get } from "../methods/index.js";
 import {
   applyObjectAssignConfigsForReactElement,
+  cloneObject,
   createInternalReactElement,
   flagPropsWithNoPartialKeyOrRef,
-  hardModifyReactObjectPropertyBinding,
   getProperty,
+  hardModifyReactObjectPropertyBinding,
   hasNoPartialKeyOrRef,
+  isEventProp,
+  transferSafePropertiesToRemoveFromObjectsToProps,
 } from "./utils.js";
 import * as t from "babel-types";
 import { computeBinary } from "../evaluators/BinaryExpression.js";
 import { CompilerDiagnostic, FatalError } from "../errors.js";
 
+function markToSafelyRemovePropertyDuringSerialization(realm: Realm, obj: ObjectValue, propName: string): void {
+  let propsToRemove = realm.react.objectsWithPropsToRemove.get(obj);
+
+  if (propsToRemove === undefined) {
+    propsToRemove = new Set();
+    realm.react.objectsWithPropsToRemove.set(obj, propsToRemove);
+  }
+  propsToRemove.add(propName);
+}
+
+function deepTraverseAndFindOrDeleteFirstRenderProperties(
+  realm: Realm,
+  obj: ObjectValue | AbstractObjectValue,
+  shouldRemoveProps: boolean,
+  alreadyVisited?: Set<AbstractObjectValue | ObjectValue> = new Set()
+): boolean {
+  if (obj instanceof FunctionValue || alreadyVisited.has(obj)) {
+    return false;
+  }
+  alreadyVisited.add(obj);
+  if (obj.constructor === ObjectValue) {
```

Comment:
Same nit about `instanceof`

Reply:
Oops, GH lost my comment. Well, it said that maybe it's better to check `instanceof AbstractObjectValue` first (if you're trying to skip it) and then check `instanceof ObjectValue`. I don't like constructor equality checks because they will break if concrete objects ever become a subclass of `ObjectValue`.

Reply:
I'll move the ordering.